### ISSUE TITLE
Missing branch coverage while using Sealed Classes

### ISSUE THREAD

**Issue by afrueh** [jacoco/jacoco#1219]
### Steps to reproduce

* JaCoCo version: 0.8.7
* Operating system: OSX
* Build Pipeline: Gradle/Android
* Minimal Example Code:

```
// SomeClass.kt
sealed class Event {
    object Foo : Event()
    data class Bar(val a: String) : Event()
}

class SomeClass {
    fun useEvent(event: Event) {
        when (event) {
            is Event.Foo -> print("Foo")
            is Event.Bar -> print("Bar ${event.a}")
        }
    }
}
```

```
//SomeClassTest.kt
import org.junit.Before
import org.junit.Test

class SomeClassTest {

    lateinit var someClass: SomeClass

    @Before
    fun setup() {
        someClass = SomeClass()
    }

    @Test
    fun event_Foo() {
        someClass.useEvent(Event.Foo)
    }

    @Test
    fun event_Bar() {
        someClass.useEvent(Event.Bar("Hello"))
    }
}

```

### Expected behaviour
Using Sealed Classes within a `when` is exhaustive. So the branch coverage should be 100% in given example.

### Actual behaviour
Jacoco reports the `else` case as an missing branch.

<img width="580" alt="Screenshot 2021-09-09 at 10 05 22" src="https://user-images.githubusercontent.com/465507/132649233-42245f74-52f7-4823-b2cd-f5106f0478d5.png">

**Comment by marqc** [comment]
+1
Same problem is observed for java 17 preview feature "Pattern matching on switch" with sealed classes

https://openjdk.java.net/jeps/406

```java
sealed interface S permits A, B, C {}
final class A implements S {}
final class B implements S {}
record C(int i) implements S {}  // Implicitly final

static int testSealedCoverage(S s) {
    return switch (s) {
        case A a -> 1;
        case B b -> 2;
        case C c -> 3;
    };
}
```

above code is perfectly valid and compiles but will report missing branch in jacoco when all 3 possibilities + NULL value is tested

**Comment by better-712** [comment]
I think the essence of the problem is that when the branch has covered all the possible cases,Jacoco shouldn't report else case as an missing branch,just like the picture linked below.
If so, I wonder whether or not it's really defined as a bug? And can i repair this bug as my first contribution to the github community
![93ECFB33-D69B-44A4-8C6D-532211328C4E](https://user-images.githubusercontent.com/89117847/159887416-897cdd33-83b0-45f8-8776-1de733d6ba5a.png)

**Comment by marchof** [comment]
@better-712 Please help us to keep the issues focused. This issue is specifically about sealed classes in Kotlin. Please feel free to to discuss your ideas on our mailing list or open a separate issue. Thx!

**Comment by better-712** [comment]
I think the bug is due to: after convert kotlin to bytecode, 'when' is almost no difference with 'if elseif'(bytecode picture is attached below).As jacoco use probe to measure branch coverage, it can’t aware the condition is exhaustive.
I'm trying to fix this bug，but it seems there is no way to deal with it in bytecode level.Could someone give me some advice?
Please don't deleted m
![B37BCD74-696A-4AEB-A633-4C3710078BD7](https://user-images.githubusercontent.com/89117847/164968274-08e6e0bd-3922-4dbf-9437-253d22bab017.png)
y comment, if i said something wrong, please tell me.I'm a CS student and want to solve some issu
![67E926F5-8C97-41CA-9F19-ABB356EDAAFB](https://user-images.githubusercontent.com/89117847/164968281-61c2b225-5b5f-4858-a474-be68328846cb.png)
![97ADDDE3-A28A-4189-8454-8582045489A8](https://user-images.githubusercontent.com/89117847/164968288-d3fbc5ef-da93-47d1-a101-ebffdcd52bf3.png)
e

**Comment by better-712** [comment]
I think the bug is due to: after convert kotlin to bytecode, 'when' is almost no difference with 'if elseif'(bytecode picture is attached below).As jacoco use probe to measure branch coverage, it can’t aware the condition is exhaustive.
I'm trying to fix this bug，but it seems there is no way to deal with it in bytecode level.Could someone give me some advice?
Please don't deleted my comment, if i said something wrong, please tell me.I'm a CS student and want to solve the issue

**Comment by better-712** [comment]
![B37BCD74-696A-4AEB-A633-4C3710078BD7](https://user-images.gi
![67E926F5-8C97-41CA-9F19-ABB356EDAAFB](https://user-images.githubusercontent.com/89117847/164968373-de32b808-4b07-4da1-9742-24e0d1544164.png)
thubusercontent.com/89117847/164968372-0f560822-d98c-431b-bcff-58a842f67aca.png)
![97ADDDE3-A28A-4189-8454-8582045489A8](https://user-images.githubusercontent.com/89117847/164968377-cd2fd629-8428-4c31-8e3e-b11312d8f692.png)

**Comment by better-712** [comment]
![B37BCD74-696A-4AEB-A633-4C3710078BD7](https://user-images.git
![67E926F5-8C97-41CA-9F19-ABB356EDAAFB](https://user-images.githubusercontent.com/89117847/164968428-3528cdd0-50c2-49db-bbf3-27235735b3bd.png)
hubusercontent.com/89117847/164968424-0f353dad-a171-46bb-9977-e513911959b2.png)
![97ADDDE3-A28A-4189-8454-8582045489A8](https://user-images.githubusercontent.com/89117847/164968430-50ab9ea6-cc18-4c84-893a-b8f3a60dfdd7.png)

**Comment by better-712** [comment]
![D26A121D-89A4-48BA-984B-B034A9842695](https://user-images.githubusercontent.com/89117847/164968495-d97270cd-2bbd-416b-94c6-44da2957fb37.png)

**Comment by better-712** [comment]
Although while using sealed class, we don't need to specify else case for when expression.After it convert to byte code, it generate an if-else if chain, no else at the end! So jacoco always consider there is a missing branch, and I can't think of any way to fix it.It seems like a compiler fault, and jacoco is not responsible for it.
Looking forward to any related reply.

**Comment by dmytroKarataiev** [comment]
It's a very annoying issue. 
We can't sacrifice compile-time checking by using an `else` branch to get 100% coverage with jacoco. Any ideas?

**Comment by joaodevsantos** [comment]
@afrueh do you have any update on this issue? I'm having a similar problem on a Sealed class that has the following structure.

`sealed class XXX<T>(val x: String? = null, val xx: T? = null)`

I have added the required unit tests however JaCoCo still complains on the following missing instruction:

`XXX(String, Object, int, DefaultConstructorMarker)`

**Comment by marvinjuette** [comment]
I maybe have a similar issue with enums. But i already posted "more" details in the following comment: https://github.com/jacoco/jacoco/issues/1295#issuecomment-1603436651.

**Comment by nick2525** [comment]
Still reproduced, any plans to fix it?

**Comment by marchof** [comment]
@nick2525 That's why the bug is "Open". Also please see our FAQ:

> **When will feature X be released?**
> 
> JaCoCo is maintained by volunteers in their free time. Since we cannot guarantee free capacity, we do not commit to particular release dates. Typically, you can expect a couple of releases every year.
> 
> In the [change log](https://www.jacoco.org/jacoco/trunk/doc/changes.html), you can see all features that have been implemented in master branch and will be available with the next release. And in the meantime you can test latest build of master branch ([Maven SNAPSHOT](https://www.jacoco.org/jacoco/trunk/doc/repo.html)) and provide [feedback](https://www.jacoco.org/jacoco/trunk/doc/support.html) to us.

**Comment by Joannanas** [comment]
Hi, I have no knowledge about this, otherwise I would've done it, but I have the same ssue with enums and JaCoCo expecting some "else" branch, and I saw that a similar issue for Java "switch" case has been fixed https://github.com/jacoco/jacoco/pull/1472 not that long ago so I let that here in case it'll be helpful to fix this one.

**Comment by nick2525** [comment]
Have you been able to reproduce this? Problem here is that many technologies, for example Sonarqube, use Jacoco and there, too, the coverage is considered incorrect.

**Comment by Joannanas** [comment]
My case seems to be a mix of both this issue and the Java "switch" fix I previously linked.

I have some class with a method doing stuff according to an enum value:
<img width="504" alt="Screenshot with code example, classic when and two cases, one per enum value, doing a basic 'print'" src="https://github.com/jacoco/jacoco/assets/73722225/88d2e197-3513-49d7-8e14-31dbe2e18a7e">

I create some test, basic one here, nothing much to do:
<img width="673" alt="Screenshot with code example of test of the previous screenshot. Two tests, one per enum value." src="https://github.com/jacoco/jacoco/assets/73722225/3d80e0bd-3d0d-4eb1-9731-0807b11b3e26">


But JaCoCo still expects some "else" case and count this as a missing branch:
![Screenshot of JaCoCo report indicating a missing branch](https://github.com/jacoco/jacoco/assets/73722225/b341ffc9-8a91-4b9f-82e0-8cbaf9c9d177)

If I actually replace my exhaustive enum with an "else" (which is obvisouly not what I need), then I get the 100% coverage:
<img width="490" alt="Screenshot with same piece of code as the first one, except that the second enum case is 'else'" src="https://github.com/jacoco/jacoco/assets/73722225/c089322c-d40e-4064-872f-74ea0a670f43">
![Screenshot 2024-06-17 at 17 42 21](https://github.com/jacoco/jacoco/assets/73722225/6cfe4c3c-dcf9-49d8-8e83-6fa564d6def5)

Also, while reproducing that to send it here, I came across a specific case when. I use "return" for both enum values (only one is not working, has to be both), and then the coverage is also 100%.
<img width="513" alt="Screenshot with the same code as the first one but doing a simple 'return' instead of 'print'" src="https://github.com/jacoco/jacoco/assets/73722225/dde3908b-54bd-464a-9594-6568aac180ff">

**Comment by Godin** [comment]
Fixed in Kotlin compiler version 2.0 - see #1791
